src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs

```diff
@@ -0,0 +1,118 @@
+
+                    List<string> requestLines = await serverTask;
+
+                    // Note since we're using a proxy, host name is included in the request line
+                    Assert.Equal($"GET http://{s_EncodedHostName}/ HTTP/1.1", requestLines[0]);
+                    Assert.Contains($"Host: {s_EncodedHostName}", requestLines);
+                }
+            });
+        }
+
+        [ActiveIssue(26355)] // We aren't doing IDNA encoding properly
+        [Fact]
+        public async Task InternationalRequestHeaderValues_UsesIdnaEncoding_Success()
+        {
+            await LoopbackServer.CreateServerAsync(async (server, url) =>
+            {
+                using (HttpClient client = new HttpClient(CreateHttpClientHandler()))
+                {
+                    var request = new HttpRequestMessage(HttpMethod.Get, url);
+                    request.Headers.Host = s_IdnaHost;
+                    request.Headers.Referrer = s_IdnaUri;
+                    Task<HttpResponseMessage> getResponseTask = client.SendAsync(request);
+                    Task<List<string>> serverTask = LoopbackServer.ReadRequestAndSendResponseAsync(server, LoopbackServer.DefaultHttpResponse);
+
+                    await TestHelper.WhenAllCompletedOrAnyFailed(getResponseTask, serverTask);
+
+                    List<string> requestLines = await serverTask;
+
+                    Assert.Contains($"Host: {s_EncodedHostName}", requestLines);
+                    Assert.Contains($"Referer: {s_EncodedUri}", requestLines);
+                }
+            });
+        }
+
+        [ActiveIssue(26355)] // We aren't doing IDNA decoding properly
+        [Fact]
+        public async Task InternationalResponseHeaderValues_UsesIdnaDecoding_Success()
+        {
+            await LoopbackServer.CreateServerAsync(async (server, url) =>
+            {
+                HttpClientHandler handler = CreateHttpClientHandler();
+                handler.AllowAutoRedirect = false;
+
+                using (HttpClient client = new HttpClient(handler))
+                {
+                    Task<HttpResponseMessage> getResponseTask = client.GetAsync(url);
+                    Task<List<string>> serverTask = LoopbackServer.ReadRequestAndSendResponseAsync(server, 
+                        $"HTTP/1.1 302 Redirect\r\nLocation: {s_EncodedUri}\r\n\r\n");
+
+                    await TestHelper.WhenAllCompletedOrAnyFailed(getResponseTask, serverTask);
+
+                    HttpResponseMessage response = await getResponseTask;
+                    Console.WriteLine($"Location header={response.Headers.Location}");
+                    Console.WriteLine($"Location header.Host={response.Headers.Location.Host}");
+                    Console.WriteLine($"Location header.IdnHost={response.Headers.Location.IdnHost}");
+                }
+            });
+        }
+
+        sealed class SimpleWebProxy : IWebProxy
```

Comment:
Don't we have one of these already somewhere? And could we just use WebProxy?

Reply:
Probably, yeah.  I'll fix.